bin/handler-sensu.rb

```diff
@@ -133,7 +133,7 @@ def parse_remediations(remediations, occurrences, severity)
   # Issue a check via the API
   def trigger_remediation(check, subscribers)
     api_request(:POST, '/request') do |req|
-      req.body = JSON.dump('check' => check, 'subscribers' => [subscribers], 'creator' => 'sensu-plugins-sensu', 'reason' => 'Auto remediation triggered')
+      req.body = JSON.dump('check' => check, 'subscribers' => subscribers, 'creator' => 'sensu-plugins-sensu', 'reason' => 'Auto remediation triggered')
```

Comment:
I am guessing based on this that subscribers must already be an array and the change makes it go from an array to an array of arrays. Can you confirm this by any chance in the logs?

Reply:
@majormoses I can't confirm it from the logs (e.g. `sensu-api.log, `sensu-server.log`), but from my additional logging step I created by opening a file and having the `trigger_remediation` method dump to it after it issues the API call. From what I've seen there, the original/problem code is an array and removing the brackets causes it to remain an array (as demonstrated by my "logging" mechanism by issuing a `subscribers.class` call that prints out in my output file). I will spin up my vagrant set up and post examples.

Reply:
I misspoke and misremembered: the `sensu-api.log` DOES show the double bracket issue, just not in the `sensu-server.log` (even with `LOG_LEVEL=debug` set) I'm not sure why that would be.

To catch what kind (i.e. class) of variable `subscribers` is, I modified `trigger_remediation` in this manner:

```
  def trigger_remediation(check, subscribers)
    output = File.open("/tmp/remediator.out", "w")
    api_request(:POST, '/request') do |req|
      req.body = JSON.dump('check' => check, 'subscribers' => [subscribers], 'creator' => 'sensu-plugins-sensu', 'reason' => 'Auto remediation triggered')
      output.puts req.body
      output.puts subscribers
      output.puts subscribers.class
    end
    output.close
  end
```
I define a file (`remediator.out`) and, after the remediation API call kicks off, I dump the contents of `req.body` to this file; along with the contents of `subscribers`. I then find out what class of variable `subscribers` is. 

Before removing the brackets (defined as the solution in this PR), here's how the various logs (`sensu-server.log`, `sensu-api.log`, and `remediator.out`) look when an event (that's using remediation) kicks off:

`sensu-server.log` does show that the request was successful, but nothing happens on the afflicted client:

```
{"timestamp":"2017-12-12T12:02:48.368761+0000","level":"info","message":"handler output","handler":{"command":"/opt/sensu/embedded/bin/handler-sensu.rb","type":"pipe","severities":["critical"],"name":"remediator"},"event":{"id":"24804bcb-cbb4-4d7e-ac70-98b1bc3a50a7"},"output":["REMEDIATION: Evaluating remediation: sensu-client-one check_ssh_proc {\"remediator_restart_sshd\"=>{\"occurrences\"=>[3, 4], \"severities\"=>[2]}} #=1 sev=2\n"]}
{"timestamp":"2017-12-12T12:03:18.367089+0000","level":"info","message":"handler output","handler":{"command":"/opt/sensu/embedded/bin/handler-sensu.rb","type":"pipe","severities":["critical"],"name":"remediator"},"event":{"id":"24804bcb-cbb4-4d7e-ac70-98b1bc3a50a7"},"output":["REMEDIATION: Evaluating remediation: sensu-client-one check_ssh_proc {\"remediator_restart_sshd\"=>{\"occurrences\"=>[3, 4], \"severities\"=>[2]}} #=2 sev=2\n"]}
{"timestamp":"2017-12-12T12:03:48.371687+0000","level":"info","message":"handler output","handler":{"command":"/opt/sensu/embedded/bin/handler-sensu.rb","type":"pipe","severities":["critical"],"name":"remediator"},"event":{"id":"24804bcb-cbb4-4d7e-ac70-98b1bc3a50a7"},"output":["REMEDIATION: Evaluating remediation: sensu-client-one check_ssh_proc {\"remediator_restart_sshd\"=>{\"occurrences\"=>[3, 4], \"severities\"=>[2]}} #=3 sev=2\nREMEDIATION: Triggering remediation check 'remediator_restart_sshd' for [\"sensu-client-one\"] [\"client:sensu-client-one\", \"sensu-client-one\"]\nREMEDIATION: Received API Response (202): {\"issued\":1513080228}, exiting.\n"]}
```

`sensu-api.log` also shows the call was successful. Here you can see the additional brackets around `subscribers`:

```
{"timestamp":"2017-12-12T12:10:48.360813+0000","level":"debug","message":"api request","request_id":"daf3b121-2369-4a2f-a83c-2bda8a8d69c8","remote_address":"127.0.0.1","user_agent":"Ruby","method":"POST","uri":"/request","query_string":null,"body":"{\"check\":\"remediator_restart_sshd\",\"subscribers\":[[\"client:sensu-client-one\",\"sensu-client-one\"]],\"creator\":\"sensu-plugins-sensu\",\"reason\":\"Auto remediation triggered\"}"}
{"timestamp":"2017-12-12T12:10:48.361309+0000","level":"info","message":"api response","request":{"request_id":"daf3b121-2369-4a2f-a83c-2bda8a8d69c8","remote_address":"127.0.0.1","user_agent":"Ruby","method":"POST","uri":"/request","query_string":null,"body":"{\"check\":\"remediator_restart_sshd\",\"subscribers\":[[\"client:sensu-client-one\",\"sensu-client-one\"]],\"creator\":\"sensu-plugins-sensu\",\"reason\":\"Auto remediation triggered\"}"},"status":202,"content_length":21,"time":0.001}
```

My `remediator.out` file shows the double brackets in the API call (the contents of `req.body`) and that the `subscribers` variable is an array:

```
[root@sensu-server ~]# cat /tmp/remediator.out
{"check":"remediator_restart_sshd","subscribers":[["client:sensu-client-one","sensu-client-one"]],"creator":"sensu-plugins-sensu","reason":"Auto remediation triggered"}
client:sensu-client-one
sensu-client-one
Array
```

Once I remove the brackets from `handler-sensu.rb` (what this PR is requesting to do) and kick off the same scenario causing a remediation event, here's how things look in the aforementioned logs:

`sensu-server.log` shows no actual change:

```
{"timestamp":"2017-12-12T12:14:18.358400+0000","level":"info","message":"handler output","handler":{"command":"/opt/sensu/embedded/bin/handler-sensu.rb","type":"pipe","severities":["critical"],"name":"remediator"},"event":{"id":"5af330ed-66d9-47c2-814e-69ed83a7f82d"},"output":["REMEDIATION: Evaluating remediation: sensu-client-one check_ssh_proc {\"remediator_restart_sshd\"=>{\"occurrences\"=>[3, 4], \"severities\"=>[2]}} #=1 sev=2\n"]}
{"timestamp":"2017-12-12T12:14:48.355052+0000","level":"info","message":"handler output","handler":{"command":"/opt/sensu/embedded/bin/handler-sensu.rb","type":"pipe","severities":["critical"],"name":"remediator"},"event":{"id":"5af330ed-66d9-47c2-814e-69ed83a7f82d"},"output":["REMEDIATION: Evaluating remediation: sensu-client-one check_ssh_proc {\"remediator_restart_sshd\"=>{\"occurrences\"=>[3, 4], \"severities\"=>[2]}} #=2 sev=2\n"]}
{"timestamp":"2017-12-12T12:15:18.359579+0000","level":"info","message":"handler output","handler":{"command":"/opt/sensu/embedded/bin/handler-sensu.rb","type":"pipe","severities":["critical"],"name":"remediator"},"event":{"id":"5af330ed-66d9-47c2-814e-69ed83a7f82d"},"output":["REMEDIATION: Evaluating remediation: sensu-client-one check_ssh_proc {\"remediator_restart_sshd\"=>{\"occurrences\"=>[3, 4], \"severities\"=>[2]}} #=3 sev=2\nREMEDIATION: Triggering remediation check 'remediator_restart_sshd' for [\"sensu-client-one\"] [\"client:sensu-client-one\", \"sensu-client-one\"]\nREMEDIATION: Received API Response (202): {\"issued\":1513080918}, exiting.\n"]}
```

`sensu-api.log` shows the additional brackets have been removed:

```
{"timestamp":"2017-12-12T12:13:18.351977+0000","level":"debug","message":"api request","request_id":"7e8bf013-570e-446c-891f-99434253819e","remote_address":"127.0.0.1","user_agent":"Ruby","method":"POST","uri":"/request","query_string":null,"body":"{\"check\":\"remediator_restart_sshd\",\"subscribers\":[\"client:sensu-client-one\",\"sensu-client-one\"],\"creator\":\"sensu-plugins-sensu\",\"reason\":\"Auto remediation triggered\"}"}
{"timestamp":"2017-12-12T12:13:18.352633+0000","level":"info","message":"api response","request":{"request_id":"7e8bf013-570e-446c-891f-99434253819e","remote_address":"127.0.0.1","user_agent":"Ruby","method":"POST","uri":"/request","query_string":null,"body":"{\"check\":\"remediator_restart_sshd\",\"subscribers\":[\"client:sensu-client-one\",\"sensu-client-one\"],\"creator\":\"sensu-plugins-sensu\",\"reason\":\"Auto remediation triggered\"}"},"status":202,"content_length":21,"time":0.001}
```

`remediator.out` shows the removal of the brackets from the API call, and that the `subscribers` variable remains an array:

```
[root@sensu-server ~]# cat /tmp/remediator.out
{"check":"remediator_restart_sshd","subscribers":["client:sensu-client-one","sensu-client-one"],"creator":"sensu-plugins-sensu","reason":"Auto remediation triggered"}
client:sensu-client-one
sensu-client-one
Array
```

All of this was tested in vagrant. My event scenario consisted of stopping `sshd` on `sensu-client-one` and tailing `/var/log/messages` to see if and when `sshd` was started (which my remediation does).

Is this helpful at all?

Reply:
It is definitely helpful, it demonstrates what is happening and that it does fix it. This confirms my theory of it currently being a nested array:
correct:
```
"subscribers":["client:sensu-client-one","sensu-client-one"],
```
incorrect:
```
"subscribers":[["client:sensu-client-one","sensu-client-one"]],
```
I am gonna take a closer look at the code to see if there is anything we can do to make this more obvious to avoid this in the future (really we need a test for this but I don't have the time) by making the `type|class` more obvious/rigid. If I don't see a quick way to make this better I will merge and release it asap.

Reply:
The best I can come up with in the short term would be to add some validation here:
https://github.com/sensu-plugins/sensu-plugins-sensu/blob/2.4.0/bin/handler-sensu.rb#L134

Something along these lines would do nicely:
```ruby
if subscribers.class != Array || subscribers.first.class != Array
  p "subscribers: #{subscribers} must be a flat array"
  exit 3
end
api_request(:POST, '/request') do |req|
      req.body = JSON.dump('check' => check, 'subscribers' => subscribers, 'creator' => 'sensu-plugins-sensu', 'reason' => 'Auto remediation triggered')
end
```

Another option is to actually perform a flatten to ensure that it is not a nested array:
```ruby
if subscribers.class != Array
  p "subscribers: #{subscribers} must be an array"
  exit 3
elsif subscribers.class == Array && subscribers.first.class != String
  subscribers.flatten
  p "subscribers: #{subscribers} must be a flat array of strings, we auto flattened: #{subscribers}"
end
api_request(:POST, '/request') do |req|
  req.body = JSON.dump('check' => check, 'subscribers' => subscribers, 'creator' => 'sensu-plugins-sensu', 'reason' => 'Auto remediation triggered')
end
```

Reply:
I guess you could probably drop off:
```
elsif subscribers.class == Array && subscribers.first.class != String
```
to be
```
elsif subscribers.first.class != String
```

As really this is covered by the `elsif` previous check rather than making the same check twice. This is more performant but arguably less easy to immediately understand. BTW I have not tested any of this code in an actual interpreter so it might need some minor tweaking.

Reply:
@majormoses I'll give it a whirl and try testing in my vagrant set up and report back soon. Thanks for the tips! I'm loving ruby but haven't gotten to the point where I "think" in it yet, haha.

Reply:
@majormoses Are we hoping the flatten stuff will eradicate the need for my original PR? I've modified my output file stuff to use the `print` method instead of `puts` (e.g. `output.print "#{subscribers}\n"`) to see the raw contents of `subscribers` and `subscribers` appears to be a flat already already; until it makes it to `api_request`. I am not sure of the intent of the brackets around `subscribers` in `req.body`. I know you didn't make that change, so you aren't privy to the reason behind it. I wanted to note that, though. I'm continuing to look at this.

Reply:
So my thought was instead of someone removing the bracket, then someone catching this breakage again, then having to troubleshoot, and then fix it again we can just do some data validation. This makes it obvious what the data is supposed to be. A simple array, because it is not so simple to test this the best thing is to provide some validation and some auto correction. I'd modify your PR to be like this:
```ruby
def trigger_remediation(check, subscribers)
  if subscribers.class != Array
     p "subscribers: #{subscribers} must be an array"
     exit 3
  elsif subscribers.first.class == Array
     subscribers.flatten
    p "subscribers: #{subscribers} must be a flat array of strings, we auto flattened: #{subscribers}"
  end
  api_request(:POST, '/request') do |req|
    req.body = JSON.dump('check' => check, 'subscribers' => [subscribers], 'creator' => 'sensu-plugins-sensu', 'reason' => 'Auto remediation triggered')
  end
end
```

While just removing the brackets fixes the code if people do things correctly it dawned on me that someone might just as easily make a mistake passing in subscribers via config or from a check or mutator so we might as well try to prevent that in 2 ways. We first want to alert the user to it so that they can correct the issue and save time troubleshooting. The other is in the case where we know the exact trap that someone falls into we can _attempt_ to fix it but still warn the user that they should update their code/config.

Reply:
We still want to remove the bracket as otherwise it does not fix the problem as after we flatten it we inflate it again.